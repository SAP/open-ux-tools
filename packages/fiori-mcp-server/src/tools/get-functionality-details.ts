import type { GetFunctionalityDetailsInput, GetFunctionalityDetailsOutput, Parameter } from '../types';
import { PageEditorApi, findByPath } from '../page-editor-api';
import type { TreeNode, PropertyPath, TreeNodeProperty } from '../page-editor-api';
import { FUNCTIONALITIES_HANDLERS } from './functionalities';
import { resolveApplication } from './utils';

export async function getFunctionalityDetails(
    params: GetFunctionalityDetailsInput
): Promise<GetFunctionalityDetailsOutput> {
    const { appPath, functionalityId } = params;
    if (!appPath) {
        throw new Error('appPath parameter is required');
    }
    if (!functionalityId) {
        throw new Error('functionalityId parameter is required');
    }
    const externalFunctionality =
        typeof functionalityId === 'string' ? FUNCTIONALITIES_HANDLERS.get(functionalityId) : undefined;
    if (externalFunctionality) {
        return externalFunctionality.getFunctionalityDetails(params);
    }
    const { pageName, propertyPath } = resolveFunctionality(functionalityId);
    const page = await getDetails(appPath, pageName);
    const details = page ? getPropertyDetails(page, propertyPath) : undefined;
    if (!details) {
        throw new Error('functionalityId was not resolved');
    }

    if (pageName) {
        details.pageName = pageName;
    }

    return details;
}

function getParameters(properties: TreeNodeProperty[]): Parameter[] {
    const parameters: Parameter[] = [];
    for (const property of properties) {
        const parameter: Parameter = {
            id: property.name,
            name: property.displayName,
            description: property.description,
            type: property.type,
            currentValue: property.value
        };
        if (property.options) {
            parameter.options = property.options.map((option) => option.key);
        }
        if (property.properties) {
            parameter.parameters = getParameters(property.properties);
        }
        parameters.push(parameter);
    }
    return parameters;
}

function getPropertyDetails(page: TreeNode, propertyPath: PropertyPath): GetFunctionalityDetailsOutput | undefined {
    const { property, node } = findByPath([page], propertyPath) ?? {};
    let details: GetFunctionalityDetailsOutput | undefined;
    if (property) {
        // Property was found by path
        const parameters = getParameters([property]);
        details = {
            id: 'change-property',
            name: 'Change property',
            // description: `Change a property. To reset or remove it, set the value to undefined. If the property's description does not specify how to disable the related feature, setting it to undefined can typically be used to disable or clear it.`,
            // There is issue in cline by applying values with undefined - throws error "Invalid JSON argument".
            // As workaround - I am using approach with null as currently there is no use case where null is real value.
            description: `Change a property. To reset, remove, or restore it to its default value, set the value to null. If the property's description does not specify how to disable the related feature, setting it to null is typically the appropriate way to disable or clear it.`,
            parameters
        };
    } else if (node && node.path.length) {
        // Node was found by path - list node properties
        let parameters: Parameter[] = [];
        for (const property of node.properties) {
            parameters = parameters.concat(getParameters([property]));
        }
        details = {
            id: 'change-property',
            name: 'Change property',
            // description: `Change a property. To reset or remove it, set the value to undefined. If the property's description does not specify how to disable the related feature, setting it to undefined can typically be used to disable or clear it.`,
            // There is issue in cline by applying values with undefined - throws error "Invalid JSON argument".
            // As workaround - I am using approach with null as currently there is no use case where null is real value.
            description: `Change a property. To reset, remove, or restore it to its default value, set the value to null. If the property's description does not specify how to disable the related feature, setting it to null is typically the appropriate way to disable or clear it.`,
            parameters
        };
    }
    return details;
}

async function getDetails(appPath: string, pageName?: string): Promise<TreeNode | undefined> {
    const project = await resolveApplication(appPath);
    if (project?.applicationAccess) {
        const pageEditorApi = new PageEditorApi(project.applicationAccess, pageName);
        return pageEditorApi.getPageTree();
    }
}

export function resolveFunctionality(functionalityId: string | string[]): {
    pageName?: string;
    propertyPath: PropertyPath;
} {
    let propertyPath: PropertyPath;
    try {
        propertyPath = typeof functionalityId === 'string' ? JSON.parse(functionalityId) : [...functionalityId];
    } catch (e) {
        throw new Error('Invalid format of functionalityId parameter');
    }
    let pageName: string | undefined = undefined;
    if (propertyPath[0] !== 'settings') {
        // Currently all application paths starts with settings and it is simple to detect pagename from path
        pageName = propertyPath.shift()?.toString();
    }

    return { pageName, propertyPath };
}
