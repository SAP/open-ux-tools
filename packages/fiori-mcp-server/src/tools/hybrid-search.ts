import { SimpleVectorService } from './services/vector-simple';
import { TextEmbeddingService } from './services/text-embedding';
import { logger } from '../utils/logger';

export type DocSearchInput = {
    query: string;
    maxResults?: number;
};

export interface SearchResultItem {
    title: string;
    category: string;
    path: string;
    score: number;
    matches: string[];
    excerpt?: string;
    content?: string;
    uri: string;
}

export interface SearchResponseData {
    query: string;
    searchType: 'hybrid' | 'limited_fallback';
    results: SearchResultItem[];
    total: number;
    error?: string;
    suggestion?: string;
}

export interface SearchContent {
    type: 'text';
    text: string;
}

export interface SearchResults {
    content: SearchContent[];
}

export type DocSearchOutput = {
    results: SearchResults;
};

/**
 * Performs hybrid search with given parameters.
 *
 * @param params The search input parameters
 * @param resultAsString Whether to return results as string format
 * @returns Promise resolving to hybrid search results
 */
export async function docSearch(
    params: DocSearchInput,
    resultAsString: boolean = false
): Promise<SearchResponseData | string> {
    const { query, maxResults } = params;

    try {
        // Initialize services
        const vectorService = new SimpleVectorService();
        const embeddingService = new TextEmbeddingService();

        await vectorService.initialize();
        await embeddingService.initialize();

        // Convert text query to embedding vector
        const queryVector = await embeddingService.generateEmbedding(query);

        // Perform semantic search with the query vector
        const searchResults = await vectorService.semanticSearch(queryVector, maxResults ?? 10);
        if (resultAsString) {
            let resultString = '';
            for (const [index, result] of searchResults.entries()) {
                resultString += `Result ${index + 1}:\n\n`;
                resultString += `${result.document.content}\n`;
                resultString += '---\n\n';
            }
            return resultString;
        } else {
            // Convert vector search results to the expected format
            return {
                query,
                searchType: 'hybrid',
                results: searchResults.map((result) => ({
                    title: result.document.title,
                    category: result.document.category,
                    path: result.document.path,
                    score: result.score,
                    matches: [], // Vector search doesn't provide specific text matches
                    excerpt: result.document.metadata?.excerpt,
                    content: result.document.content,
                    uri: `sap-fiori://docs/${result.document.category}/${result.document.id}`
                })),
                total: searchResults.length
            };
        }
    } catch (error) {
        // Fallback when embeddings data is not available
        // Log warning about embeddings not being available
        // console.warn('Embeddings data not available, providing limited search capability:', error);
        logger.warn(`Embeddings data not available, providing limited search capability: ${error}`);
        return {
            query,
            searchType: 'limited_fallback',
            error: 'Embeddings data not available. Please install @sap-ux/fiori-docs-embeddings for full search capabilities.',
            results: [],
            total: 0,
            suggestion: 'Try running: npm install -g @sap-ux/fiori-docs-embeddings'
        };
    }
}
