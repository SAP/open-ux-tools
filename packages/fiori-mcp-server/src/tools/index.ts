import type { Tool } from '@modelcontextprotocol/sdk/types.js';
import * as Input from '../types/input';
import * as Output from '../types/output';
import { convertToSchema } from './utils';

export { docSearch } from './hybrid-search';
export type { SearchResponseData } from './hybrid-search';
export { listFioriApps } from './list-fiori-apps';
export { listFunctionalities } from './list-functionalities';
export { getFunctionalityDetails } from './get-functionality-details';
export { executeFunctionality } from './execute-functionality';

export const tools = [
    {
        name: 'search_docs',
        title: 'Search in Fiori Documentation',
        description:
            "Searches code snippets of Fiori Elements, Annotations, SAPUI5, Fiori tools documentation for the given query. You MUST use this tool if you're unsure about Fiori APIs. Optionally returns only code blocks.",
        inputSchema: {
            type: 'object',
            properties: {
                query: {
                    type: 'string',
                    description: 'Search query'
                },
                maxResults: {
                    type: 'number',
                    description: 'Maximum number of results to return',
                    default: 25
                }
            },
            required: ['query']
        }
    },
    {
        name: 'list_fiori_apps',
        description: `Scans a specified directory to find existing SAP Fiori applications that can be modified.
                    This is an optional, preliminary tool.
                    **Use this first ONLY if the target application's name or path is not already known.**
                    The output can be used to ask the user for clarification before starting the main 3-step workflow.`,
        inputSchema: convertToSchema(Input.ListFioriAppsInputSchema),
        outputSchema: convertToSchema(Output.ListFioriAppsOutputSchema)
    },
    {
        name: 'list_functionality',
        description: `**(Step 1 of 3)**
                    Gets the complete and exclusive list of supported functionalities to create a new or modify an existing SAP Fiori application.
                    This is the **first mandatory step** to begin the workflow and requires a valid absolute path to a SAP Fiori application as input.
                    You MUST use a functionalityId from this tool's output to request details to the functionality in 'get_functionality_details' (Step 2).
                    You MUST not use a functionalityId as name of a tool.
                    Do not guess, assume, or use any functionality not present in this list, as it is invalid and will cause the operation to fail.
                    **Note: If the target application is not known, use the list_fiori_apps tool first to identify it.**
                    if the functionality list does not include a functionality to support the current goal, then try using the search_docs tool as a fallback.`,
        inputSchema: convertToSchema(Input.ListFunctionalitiesInputSchema),
        outputSchema: convertToSchema(Output.ListFunctionalitiesOutputSchema)
    },
    {
        name: 'get_functionality_details',
        description: `**(Step 2 of 3)**
                    Gets the required parameters and detailed information for a specific functionality to create a new or modify an existing SAP Fiori application.
                    You MUST provide a functionalityId obtained from 'list_functionality' (Step 1).
                    The output of this tool is required for the final step 'execute_functionality' (Step 3).`,
        inputSchema: convertToSchema(Input.GetFunctionalityDetailsInputSchema),
        outputSchema: convertToSchema(Output.GetFunctionalityDetailsOutputSchema)
    },
    {
        name: 'execute_functionality',
        description: `**(Step 3 of 3)**
                    Executes a specific functionality to create a new or modify an existing SAP Fiori application with provided parameters.
                    This is the **final step** of the workflow and performs the actual creation or modification.
                    You MUST provide the exact parameter information obtained from get_functionality_details (Step 2).`,
        inputSchema: convertToSchema(Input.ExecuteFunctionalityInputSchema),
        outputSchema: convertToSchema(Output.ExecuteFunctionalityOutputSchema)
    }
] as Tool[];
