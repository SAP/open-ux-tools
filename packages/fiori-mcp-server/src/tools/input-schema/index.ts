export { default as executeFunctionalityInputSchema } from './execute-functionality.json';
export { default as getFunctionalityDetailsInputSchema } from './get-functionality-details.json';
export { default as listFioriAppsInputSchema } from './list-fiori-apps.json';
export { default as listFunctionalityInputSchema } from './list-functionality.json';
