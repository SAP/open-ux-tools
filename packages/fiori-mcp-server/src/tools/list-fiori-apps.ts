import { findFioriArtifacts } from '@sap-ux/project-access';
import type { AllAppResults } from '@sap-ux/project-access';
import type { FioriApp, ListFioriAppsInput, ListFioriAppsOutput } from '../types';

export async function listFioriApps(params: ListFioriAppsInput): Promise<ListFioriAppsOutput> {
    const { searchPath = [] } = params;
    const newFoundFioriArtifacts = await findFioriArtifacts({
        wsFolders: searchPath,
        artifacts: ['applications']
    });

    return {
        applications:
            newFoundFioriArtifacts.applications?.map((app: AllAppResults): FioriApp => {
                return {
                    name: app.manifest['sap.app']?.id,
                    path: app.appRoot,
                    type: 'list-report',
                    version: app.manifest['sap.app']?.dataSources?.mainService?.settings?.odataVersion ?? '4.0'
                };
            }) ?? []
    };
}
