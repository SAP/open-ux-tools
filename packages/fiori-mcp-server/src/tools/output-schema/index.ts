export { default as executeFunctionalityOutputSchema } from './execute-functionality.json';
export { default as getFunctionalityDetailsOutputSchema } from './get-functionality-details.json';
export { default as listFioriAppsOutputSchema } from './list-fiori-apps.json';
export { default as listFunctionalityOutputSchema } from './list-functionality.json';
