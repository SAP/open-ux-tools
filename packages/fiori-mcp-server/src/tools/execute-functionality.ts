import type {
    ExecuteFunctionalitiesInput,
    ExecuteFunctionalityOutput,
    GetFunctionalityDetailsOutput,
    Parameter
} from '../types';
import { getFunctionalityDetails, resolveFunctionality } from './get-functionality-details';
import type { PropertyPath } from '../page-editor-api';
import { PageEditorApi } from '../page-editor-api';
import { FUNCTIONALITIES_HANDLERS } from './functionalities';
import { resolveApplication } from './utils';

/**
 * Executes a functionality based on the provided parameters.
 *
 * @param params - Input parameters for executing the functionality
 * @returns A promise that resolves to the execution output
 * @throws Error if required parameters are missing
 */
export async function executeFunctionality(params: ExecuteFunctionalitiesInput): Promise<ExecuteFunctionalityOutput> {
    const { functionalityId, parameters, appPath } = params;
    if (!functionalityId) {
        throw new Error('functionalityId parameter is required');
    }
    if (!appPath) {
        throw new Error('appPath parameter is required');
    }
    // Custom/external functionalities
    const externalFunctionality =
        typeof functionalityId === 'string' ? FUNCTIONALITIES_HANDLERS.get(functionalityId) : undefined;
    if (externalFunctionality) {
        return externalFunctionality.executeFunctionality(params);
    }

    // Get functionality details to validate parameters
    const functionality = await getFunctionalityDetails({
        appPath,
        functionalityId
    });

    // Validate required parameters
    const missingParams = functionality.parameters
        .filter((param) => param.required && !(param.id in parameters))
        .map((param) => param.name);

    if (missingParams.length > 0) {
        throw new Error(`Missing required parameters: ${missingParams.join(', ')}`);
    }

    const changes = await generateChanges(functionality, functionalityId, parameters, appPath, functionality.pageName);
    let executionResult: ExecuteFunctionalityOutput;
    if (changes.length) {
        // Mock execution - in a real implementation, this would modify the actual files
        executionResult = {
            functionalityId,
            status: 'success',
            message: `Successfully executed '${functionality.name}'`,
            parameters,
            appPath,
            changes,
            timestamp: new Date().toISOString()
        };
    } else {
        // No changes register
        executionResult = {
            functionalityId,
            status: 'skipped',
            message: `Execution of '${functionality.name}' did not trigger any change in files`,
            parameters,
            appPath,
            changes,
            timestamp: new Date().toISOString()
        };
    }

    return executionResult;
}

/**
 * Generates changes based on the functionality and parameters.
 *
 * @param functionality - Details of the functionality to be executed
 * @param functionalityId - ID of the functionality
 * @param parametersValue - Values for the parameters
 * @param appPath - Path to the application
 * @param pageName - Optional name of the page
 * @returns A promise that resolves to an array of change descriptions
 */
async function generateChanges(
    functionality: GetFunctionalityDetailsOutput,
    functionalityId: string | string[],
    parametersValue: { [key: string]: unknown },
    appPath: string,
    pageName?: string
): Promise<string[]> {
    const changes: string[] = [];
    const editor = await getEditorApi(appPath, pageName);
    if (!editor) {
        return [];
    }
    const { propertyPath } = resolveFunctionality(functionalityId);
    const changedParameterInfo = findParameterById(functionality, propertyPath[propertyPath.length - 1]);

    let changed = false;
    if (!changedParameterInfo && typeof parametersValue === 'object') {
        // Parameters most likely in node parameters - edge case
        for (const parameterValue in parametersValue) {
            const paramPropertyPath = [...propertyPath, parameterValue];
            const parameterInfo = findParameterById(functionality, parameterValue);
            if (parameterInfo) {
                await editor.changeProperty(
                    paramPropertyPath,
                    resolveParameterValue(paramPropertyPath, parametersValue, parameterInfo)
                );
                changed = true;
                if (changes.length === 0) {
                    changes.push('Modified webapp/manifest.json');
                }
            }
        }
    }

    if (!changed) {
        // Common way to change property - AI passes precise property id and parameters
        await editor.changeProperty(
            propertyPath,
            resolveParameterValue(propertyPath, parametersValue, changedParameterInfo)
        );
        // problem -> result?.manifestChangeIndicator does not return changed indicator when we change fcl
        changes.push('Modified webapp/manifest.json');
    }

    return changes;
}

/**
 * Retrieves the PageEditorApi instance for the given application path and page name.
 *
 * @param appPath - Path to the application
 * @param pageName - Optional name of the page
 * @returns A promise that resolves to a PageEditorApi instance or undefined
 */
export async function getEditorApi(appPath: string, pageName?: string): Promise<PageEditorApi | undefined> {
    const project = await resolveApplication(appPath);
    if (project?.applicationAccess) {
        return new PageEditorApi(project.applicationAccess, pageName);
    }
    return undefined;
}

/**
 * Currently resolved values through params passed differently time to time by AI.
 * This method tries to resolve value before applying/saving it.
 *
 * @param propertyPath - Path to the property
 * @param parametersValue - Object containing parameter values
 * @param parameterInfo - Optional information about the parameter
 * @returns The resolved parameter value
 */
function resolveParameterValue(
    propertyPath: PropertyPath,
    parametersValue: { [key: string]: unknown },
    parameterInfo?: Parameter
): unknown {
    const propertyName = propertyPath[propertyPath.length - 1];
    if (parameterInfo?.type === 'object') {
        let value: unknown = parametersValue;
        if (propertyName in parametersValue) {
            // Change property is part of parameters object
            value = parametersValue[propertyName];
        }
        if (value === null || (typeof value === 'object' && !Array.isArray(value) && Object.keys(value).length === 0)) {
            // AI fails to pass undefined value, it is more stablew with null - currently workaround is to convert it to 'undefined'
            // So far there is no valid property with value null - such transformation seems ok
            value = undefined;
        }
        // Whole object is received as parameters
        return value;
    }
    let value;
    if (propertyName in parametersValue) {
        // Change property is part of parameters object
        value = parametersValue[propertyName];
    } else if ('value' in parametersValue) {
        // Seems generic approach from AI - property named 'value' contains new value
        value = parametersValue.value;
    }
    if (value === null) {
        // AI fails to pass undefined value, it is more stablew with null - currently workaround is to convert it to 'undefined'
        // So far there is no valid property with value null - such transformation seems ok
        value = undefined;
    }
    return value;
}

/**
 * Finds a parameter in the functionality details by its ID.
 *
 * @param functionality - Details of the functionality
 * @param id - ID of the parameter to find
 * @returns The found Parameter object or undefined if not found
 */
function findParameterById(functionality: GetFunctionalityDetailsOutput, id?: string | number): Parameter | undefined {
    return functionality.parameters.find((parameter) => parameter.id === id);
}
