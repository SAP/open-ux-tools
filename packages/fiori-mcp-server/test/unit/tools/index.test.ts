import { tools } from '../../../src/tools/index';

const listFioriApps = tools.find((tool) => tool.name === 'list_fiori_apps');
const listFunctionality = tools.find((tool) => tool.name === 'list_functionality');
const getFunctionalityDetails = tools.find((tool) => tool.name === 'get_functionality_details');
const executeFunctionaliy = tools.find((tool) => tool.name === 'execute_functionality');
const getFioriRules = tools.find((tool) => tool.name === 'get_fiori_rules');

describe('Tools schemas', () => {
    test('list_fiori_apps', async () => {
        expect(listFioriApps?.inputSchema).toMatchSnapshot('Input schema for "list_fiori_apps"');
        expect(listFioriApps?.outputSchema).toMatchSnapshot('Output schema for "list_fiori_apps"');
    });

    test('list_functionality', async () => {
        expect(listFunctionality?.inputSchema).toMatchSnapshot('Input schema for "list_functionality"');
        expect(listFunctionality?.outputSchema).toMatchSnapshot('Output schema for "list_functionality"');
    });

    test('get_functionality_details', async () => {
        expect(getFunctionalityDetails?.inputSchema).toMatchSnapshot('Input schema for "get_functionality_details"');
        expect(getFunctionalityDetails?.outputSchema).toMatchSnapshot('Output schema for "get_functionality_details"');
    });

    test('execute_functionality', async () => {
        expect(executeFunctionaliy?.inputSchema).toMatchSnapshot('Input schema for "execute_functionality"');
        expect(executeFunctionaliy?.outputSchema).toMatchSnapshot('Output schema for "execute_functionality"');
    });

    test('get_fiori_rules', async () => {
        expect(getFioriRules).toBeDefined();
        expect(getFioriRules?.name).toBe('get_fiori_rules');
        expect(getFioriRules?.description).toContain('rules and best practices');
        expect(getFioriRules?.inputSchema).toBeDefined();
        expect(getFioriRules?.inputSchema.type).toBe('object');
        expect(getFioriRules?.inputSchema.properties).toEqual({});
        expect(getFioriRules?.inputSchema.required).toEqual([]);
    });
});
