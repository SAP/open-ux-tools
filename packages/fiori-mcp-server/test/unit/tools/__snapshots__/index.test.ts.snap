// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Tools schemas execute_functionality: Input schema for "execute_functionality" 1`] = `
Object {
  "additionalProperties": false,
  "description": "Input object for executing a functionality. Only three top-level properties are allowed: \\"functionalityId\\", \\"parameters\\", and \\"appPath\\". All other dynamic or functionality-specific inputs must be included inside the \\"parameters\\" object. Do not place any additional fields at the root level.",
  "properties": Object {
    "appPath": Object {
      "description": "Path to the Fiori application. Path should be an absolute path.",
      "type": "string",
    },
    "functionalityId": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "items": Object {
            "anyOf": Array [
              Object {
                "type": "string",
              },
              Object {
                "type": "number",
              },
            ],
          },
          "type": "array",
        },
      ],
      "description": "The ID of the functionality to execute",
    },
    "parameters": Object {
      "additionalProperties": Object {},
      "description": "Parameters for the functionality execution",
      "propertyNames": Object {
        "type": "string",
      },
      "type": "object",
    },
  },
  "required": Array [
    "functionalityId",
    "parameters",
    "appPath",
  ],
  "type": "object",
}
`;

exports[`Tools schemas execute_functionality: Output schema for "execute_functionality" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "appPath": Object {
      "type": "string",
    },
    "changes": Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "functionalityId": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "items": Object {
            "anyOf": Array [
              Object {
                "type": "string",
              },
              Object {
                "type": "number",
              },
            ],
          },
          "type": "array",
        },
      ],
    },
    "message": Object {
      "type": "string",
    },
    "parameters": Object {},
    "status": Object {
      "type": "string",
    },
    "timestamp": Object {
      "type": "string",
    },
  },
  "required": Array [
    "functionalityId",
    "status",
    "message",
    "parameters",
    "appPath",
    "changes",
    "timestamp",
  ],
  "type": "object",
}
`;

exports[`Tools schemas get_functionality_details: Input schema for "get_functionality_details" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "appPath": Object {
      "description": "Path to the Fiori application. Path should be an absolute path.",
      "type": "string",
    },
    "functionalityId": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "items": Object {
            "anyOf": Array [
              Object {
                "type": "string",
              },
              Object {
                "type": "number",
              },
            ],
          },
          "type": "array",
        },
      ],
      "description": "The ID of the functionality to get details for",
    },
  },
  "required": Array [
    "appPath",
    "functionalityId",
  ],
  "type": "object",
}
`;

exports[`Tools schemas get_functionality_details: Output schema for "get_functionality_details" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "description": Object {
      "type": "string",
    },
    "examples": Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "functionalityId": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "items": Object {
            "anyOf": Array [
              Object {
                "type": "string",
              },
              Object {
                "type": "number",
              },
            ],
          },
          "type": "array",
        },
      ],
      "description": "Identifier to pass as the \`functionalityId\` parameter when calling \`get_functionality_details\` or \`execute_functionality\`",
    },
    "impact": Object {
      "type": "string",
    },
    "name": Object {
      "type": "string",
    },
    "pageName": Object {
      "type": "string",
    },
    "parameters": Object {
      "additionalProperties": Object {},
      "properties": Object {
        "description": Object {
          "type": "string",
        },
        "properties": Object {
          "additionalProperties": Object {
            "anyOf": Array [
              Object {
                "additionalProperties": Object {},
                "properties": Object {
                  "description": Object {
                    "type": "string",
                  },
                  "properties": Object {},
                  "type": Object {
                    "anyOf": Array [
                      Object {
                        "enum": Array [
                          "object",
                          "array",
                          "string",
                          "number",
                          "boolean",
                          "null",
                          "integer",
                          "any",
                        ],
                        "type": "string",
                      },
                      Object {
                        "items": Object {
                          "enum": Array [
                            "object",
                            "array",
                            "string",
                            "number",
                            "boolean",
                            "null",
                            "integer",
                            "any",
                          ],
                          "type": "string",
                        },
                        "type": "array",
                      },
                    ],
                  },
                },
                "type": "object",
              },
              Object {
                "type": "boolean",
              },
            ],
          },
          "propertyNames": Object {
            "type": "string",
          },
          "type": "object",
        },
        "type": Object {
          "anyOf": Array [
            Object {
              "enum": Array [
                "object",
                "array",
                "string",
                "number",
                "boolean",
                "null",
                "integer",
                "any",
              ],
              "type": "string",
            },
            Object {
              "items": Object {
                "enum": Array [
                  "object",
                  "array",
                  "string",
                  "number",
                  "boolean",
                  "null",
                  "integer",
                  "any",
                ],
                "type": "string",
              },
              "type": "array",
            },
          ],
        },
      },
      "type": "object",
    },
    "prerequisites": Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "technicalDescription": Object {
      "type": "string",
    },
  },
  "required": Array [
    "functionalityId",
    "name",
    "description",
    "parameters",
  ],
  "type": "object",
}
`;

exports[`Tools schemas list_fiori_apps: Input schema for "list_fiori_apps" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "searchPath": Object {
      "description": "Path to search for Fiori applications (defaults to current directory). If VSCode - list of VS Code workspace folder paths(\`workspace.workspaceFolders\`)",
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
  },
  "required": Array [
    "searchPath",
  ],
  "type": "object",
}
`;

exports[`Tools schemas list_fiori_apps: Output schema for "list_fiori_apps" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "applications": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "appPath": Object {
            "description": "Absolute path to the Fiori application's root directory.",
            "type": "string",
          },
          "name": Object {
            "description": "Name of the Fiori application. Usually derived from the \`sap.app/id\` field in the application's manifest.json.",
            "type": "string",
          },
          "odataVersion": Object {
            "description": "OData protocol version used by the application's main service.",
            "type": "string",
          },
          "projectPath": Object {
            "description": "Absolute path to the root directory of the project containing this Fiori application. For EDMXBackend (standalone) projects, this is the same as \`appPath\`. For CAP projects, this points to the CAP project root, which may contain multiple Fiori applications.",
            "type": "string",
          },
          "projectType": Object {
            "description": "Type of project the application belongs to.",
            "enum": Array [
              "EDMXBackend",
              "CAPJava",
              "CAPNodejs",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "appPath",
          "projectPath",
          "projectType",
          "odataVersion",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [
    "applications",
  ],
  "type": "object",
}
`;

exports[`Tools schemas list_functionality: Input schema for "list_functionality" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "appPath": Object {
      "description": "Path to the root folder of the Fiori application (where package.json and ui5.yaml reside). Path should be an absolute path.",
      "type": "string",
    },
  },
  "required": Array [
    "appPath",
  ],
  "type": "object",
}
`;

exports[`Tools schemas list_functionality: Output schema for "list_functionality" 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "applicationPath": Object {
      "type": "string",
    },
    "functionalities": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "description": Object {
            "type": "string",
          },
          "functionalityId": Object {
            "anyOf": Array [
              Object {
                "type": "string",
              },
              Object {
                "items": Object {
                  "anyOf": Array [
                    Object {
                      "type": "string",
                    },
                    Object {
                      "type": "number",
                    },
                  ],
                },
                "type": "array",
              },
            ],
            "description": "Identifier to pass as the \`functionalityId\` parameter when calling \`get_functionality_details\` or \`execute_functionality\`",
          },
        },
        "required": Array [
          "functionalityId",
          "description",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [
    "applicationPath",
    "functionalities",
  ],
  "type": "object",
}
`;
